### ISSUE TITLE
Implement a disk replacement UI

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#1611]
At times it is desired to replace an existing disk with another one, ie the function of:

    btrfs replace start devid /dev/sdX /mnt2/pool-name

A user level interface for this process would make for a nice improvement. We should ensure to allow for the -r switch where the source device is considered read only (recommended for when there are many read errors on that device). The status of this or other current replace operations could be shown in a tab on the same page, akin to the balance tab in pools maybe.
The current status can be had via:

    btrfs replace status /mnt2/pool-name

N.B. it is generally considered to be a longer process to use replace rather than:
"btrfs dev add" and then "btrfs dev delete", it might make sense to suggest this course of action in the same UI.

**Comment by phillxnet** [comment]
Please also update the following forum thread with development in this area:
https://forum.rockstor.com/t/bad-drive-add-new-drive-problem/2798

**Comment by phillxnet** [comment]
Linking to prior duplicate issue. Suggest we maintain this issue as prior issue has less info and unsure of "...much faster..." claim.
Linking to #593
https://github.com/rockstor/rockstor-core/issues/593

**Comment by phillxnet** [comment]
Please also update the following forum thread with this issues resolution:
https://forum.rockstor.com/t/defective-disk-any-recommendations-regarding-replacement/3037

**Comment by phillxnet** [comment]
Please also update the following forum thread with this issues resolution:
https://forum.rockstor.com/t/btrfs-replace-and-possibility-of-custom-dashboard-widgets/3775

**Comment by phillxnet** [comment]
Linking to another more recent forum thread concerning the utility of a replace Web-UI element:
https://forum.rockstor.com/t/replacing-disks-in-a-raid10-pool/5130

**Comment by phillxnet** [comment]
It is proposed that we extend the existing 'pool_resize_wizard' to include this feature.

While working on the related issue:
"Implement a delete missing disk in pool UI" #1700
the following UI entry point is to be added as a place holder:

![proposed-disk-replace-ui-entry-point](https://user-images.githubusercontent.com/2521585/46344376-d5121380-c638-11e8-9106-bc825777c434.png)

This should help with those looking for this feature within the UI and identify it, for the time being, as a 'planned option'.

**Comment by Hooverdan96** [comment]
@phillxnet if I am not mistaken this can be closed now? Or do we have any outstanding UI work for this particular piece?

**Comment by lxwulf** [comment]
When can we count approximately with the release of this function? Else great work! 😊

**Comment by phillxnet** [comment]
@lxwulf We are now late in our current testing phase and so this will not be approached until the next testing phase: but the next testing phase is to be focused on some code tidy-ups ready for our over-due front-end updates (for the subsequent testing branch phase). It may be this disk replacement capability will be approached then: but for the time being we have the existing disk removal/addition mechanism: two steps, with less efficiency, but entirely practical in most settings.

Apologies for not being able to pin this down time-wise for the time being at least. But it is still very much in mind as a future capability: hence the greyed-out placeholder. Thanks for your kind words.

**Comment by lxwulf** [comment]
Hello @phillxnet :wave: ,

That's a bummer, but I understand the sometimes chaotic or better say overwhelming to-do list in such a big project like this one here - so no problem there, at least we have a function to do it. 😊

Just to be sure the steps are following:

- Start removing process for the failure disk
- Remove the disk physically
- Add the new disk physically
- Start the adding process for the new disk
- done

Many thanks for your detailed and fast response and excuse for my very slow response, nonetheless I really appreciate that!

**Comment by phillxnet** [comment]
@lxwulf Yes, always lots to do. And it may be we get a Pull Request contribution seeing to this planned feature sooner than the regular contributors are able to. So keep an eye on the change-log.

Regarding your proposed steps. Yes, that is one way. You could also, if need be, add a disk to the pool first: to avoid the Pool dropping below the Pool btrfs-raid lower disk-count limit. Or change the raid level to accommodate fewer disks. Both have risks as they can entail further stress on an already failing drive, and reduce (or remove) redundancy. Or you could use the suggested CLI command `-r` option:

> We should ensure to allow for the -r switch where reads from the source device are only carried out if not other zero-defect mirror exists (recommended for poorly devices). See: https://btrfs.readthedocs.io/en/latest/btrfs-replace.html

that we propose to introduce in this issue, i.e. treat the to-be-replaced dive as a last resort for data retrieval. Quite a few options really - and the best choice is difficult to discern, but depends in-part on how poorly the failing drive actually is.

We also have the following guide which can help with such things:

https://rockstor.com/docs/data_loss.html

As always, before doing anything significant, ensure your back-ups are fresh. Adding, removing, replacing Pool members, or changing raid-levels, are non-trivial events. So don't have the systems associated data be the only copy that exists.

This type of discussion is better located on our forum: https://forum.rockstor.com/ so do start a fresh thread with your particular setup details (Pool raid level, number of members, Rockstor version etc) if you fancy mulling over your available options further. There is a lot more experience there than I can offer here in the GitHub issues section. And do give details on the level of poorly-ness of the drive you are trying to replace. One can also deal with missing drives: but only if the remaining drives are sufficient in number as to not be below the associated raid limit.

Hope that helps.

**Comment by lxwulf** [comment]
@phillxnet many thanks again for this big and detailed insight! 🤝

I saved your links to be read and also created a forum account for possible further help. Hopefully I can also give something back soon. 😊
